# **Data Understanding**

This notebook loads the e-commerce dataset and inspects its structure, data types, summary statistics, missing values, and duplicate records.

In [2]:
# Load pandas for data manipulation.
import pandas as pd

# Set option to display all columns and format float values to two decimal places.
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:,.2f}'.format)

In [3]:
# Read the raw e-commerce dataset.
ecommerce_df = pd.read_csv(r'dataset/raw/ecommerce_dataset_+1m.csv')

## 1. Look at the Data

In [4]:
# Display a random sample of 10 rows
ecommerce_df.sample(10)

,order_id,order_date,order_year,order_month,order_day,order_hour,order_minute,order_second,is_weekend,order_status,return_reason,customer_id,customer_name,gender,age,customer_segment,country,city,customer_loyalty_score,total_orders_by_customer,account_creation_date,product_id,product_name,category,sub_category,brand,product_rating_avg,product_reviews_count,stock_quantity,unit_price_usd,quantity,discount_percent,discount_amount_usd,total_price_usd,cost_usd,profit_usd,tax_usd,currency,payment_method,payment_status,installment_plan,shipping_method,shipping_cost_usd,delivery_days,shipping_country,warehouse_location,delivery_status,rating,review_sentiment,customer_feedback,coupon_used,coupon_code,campaign_source,device_type,traffic_source,session_duration_minutes,pages_visited,abandoned_cart_before,fraud_risk_score,profit_margin_percent,order_priority,support_ticket_created
458660,ORD-U4B8M,2024-11-05 14:49:07.354723,2024,11,5,14,49,7,No,Completed,NaN,CUS-2C30H,Nicholas Wilkerson,Male,44,Regular,Canada,Toronto,91.60,27,2021-09-26,PRD-A44M,Running Shoes Air Zoom,Clothing,Mens Wear,Canon,4.30,1784,604,147.13,4,0,0.00,588.52,292.64,295.88,47.35,USD,Debit Card,Paid,No,Standard,17.03,14,Canada,EU-Central,Delivered,4,Positive,Excellent!,Yes,CPN-4IG,Organic,Desktop,Email,57.60,15,No,60.20,50.28,High,No
645482,ORD-QQWZF,2025-05-08 02:12:51.345662,2025,5,8,2,12,51,No,Completed,NaN,CUS-AXIQP,Jordan White,Female,50,Premium,Spain,Bilbao,4.70,10,2024-02-09,PRD-CUCY,Coffee Maker Automatic,Home,Decor,Apple,3.10,604,594,246.27,5,0,0.00,"1,231.35",530.35,701.00,63.83,USD,PayPal,Paid,No,Express,14.78,12,Spain,USA-East,In Transit,3,Neutral,Not satisfied,No,NaN,Email,Desktop,Search,10.60,20,No,95.80,56.93,Medium,Yes
404598,ORD-XMWME,2025-01-12 20:24:23.696458,2025,1,12,20,24,23,Yes,Completed,NaN,CUS-MFAS1,Pamela Young,Female,24,Premium,Belgium,Bruges,27.30,35,2021-12-30,PRD-NAOO,Resistance Bands Set,Health,Personal Care,Logitech,4.30,4532,6,46.87,4,0,0.00,187.48,97.24,90.24,16.88,USD,Debit Card,Paid,Yes,Standard,7.23,7,Belgium,UK,Delivered,5,Positive,Could be better,Yes,CPN-84P,Affiliate,Desktop,Social,44.70,4,Yes,34.80,48.13,Medium,No
82999,ORD-ZEPVW,2025-02-03 00:08:08.983692,2025,2,3,0,8,8,No,Completed,NaN,CUS-TO7A0,James Higgins,Male,54,VIP,Belgium,Ghent,38.60,25,2021-09-16,PRD-26M8,Digital Thermometer,Health,Fitness,Fitbit,4.50,4245,914,128.63,5,0,0.00,643.15,317.65,325.50,47.66,USD,Credit Card,Paid,Yes,Express,2.84,7,Belgium,UK,Failed,4,Positive,Good value for money,No,NaN,Facebook,Desktop,Direct,10.70,15,No,12.70,50.61,High,Yes
228416,ORD-DLXV8,2024-09-22 09:03:05.411916,2024,9,22,9,3,5,Yes,Completed,NaN,CUS-A95F3,Christopher Smith,Male,52,Regular,Netherlands,Eindhoven,75.90,47,2022-02-12,PRD-3D4C,Exercise Ball,Sports,Accessories,Baseus,3.50,3235,226,134.43,3,15,60.49,342.80,167.19,175.61,20.06,USD,Bank Transfer,Paid,Yes,Economy,12.36,13,Netherlands,USA-East,Delivered,5,Positive,Average quality,No,NaN,Organic,Tablet,Direct,47.70,4,Yes,95.20,51.23,Low,No
631440,ORD-LSHCN,2025-12-12 12:01:14.389127,2025,12,12,12,1,14,No,Pending,NaN,CUS-RSOX3,Caleb Hill,Male,29,Regular,United States,Houston,75.00,23,2022-01-30,PRD-CND7,Air Fryer 5L,Home,Bedding,Fitbit,4.20,2349,223,126.00,5,0,0.00,630.00,399.30,230.70,50.62,USD,Debit Card,Paid,Yes,Express,3.35,2,United States,USA-East,Delivered,5,Positive,NaN,Yes,CPN-NBR,Facebook,Desktop,Email,45.10,8,Yes,1.50,36.62,Low,Yes
125279,ORD-BN7AR,2024-10-29 22:53:03.475707,2024,10,29,22,53,3,No,Returned,Not As Described,CUS-0ZQ3N,Brendan Vaughan,Male,23,Premium,Italy,Turin,70.10,24,2022-06-25,PRD-UFUN,Winter Jacket Waterproof,Clothing,Kids Wear,Xiaomi,5.00,3691,17,146.71,1,15,22.01,124.70,100.96,23.74,7.60,USD,Debit Card,Paid,No,Express,14.72,11,Italy,UK,Delivered,1,Negative,NaN,Yes,CPN-H0U,Organic,Tablet,Direct,5.90,10,Yes,78.90,19.04,Medium,Yes
126396,ORD-OPJP7,2024-08-07 09:03:04.867820,2024,8,7,9,3,4,No,Completed,NaN,CUS-E2E0Y,Harold Guerrero,Male,37,Premium,United States,Chicago,28.70,39,2023-11-21,PRD-H1TS,Fitness

In [5]:
# Dataset dimensions (rows, columns)
ecommerce_df.shape

(1000123, 62)

## 2. Data Types & Structure

In [6]:
# Concise summary of column types and non-null counts
ecommerce_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000123 entries, 0 to 1000122
Data columns (total 62 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   order_id                  1000123 non-null  object 
 1   order_date                1000123 non-null  object 
 2   order_year                1000123 non-null  int64  
 3   order_month               1000123 non-null  int64  
 4   order_day                 1000123 non-null  int64  
 5   order_hour                1000123 non-null  int64  
 6   order_minute              1000123 non-null  int64  
 7   order_second              1000123 non-null  int64  
 8   is_weekend                1000123 non-null  object 
 9   order_status              1000123 non-null  object 
 10  return_reason             99879 non-null    object 
 11  customer_id               1000123 non-null  object 
 12  customer_name             1000123 non-null  object 
 13  gender                    1

## 3. Descriptive Statistics

In [7]:
# Summary statistics for numerical columns
ecommerce_df.describe()

,order_year,order_month,order_day,order_hour,order_minute,order_second,age,customer_loyalty_score,total_orders_by_customer,product_rating_avg,product_reviews_count,stock_quantity,unit_price_usd,quantity,discount_percent,discount_amount_usd,total_price_usd,cost_usd,profit_usd,tax_usd,shipping_cost_usd,delivery_days,rating,session_duration_minutes,pages_visited,fraud_risk_score,profit_margin_percent
count,"1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00"
mean,"2,024.59",6.51,15.75,11.49,29.49,29.49,46.49,49.97,25.49,4.00,"2,501.57",499.81,147.02,3.00,8.50,37.48,403.24,242.31,160.93,30.25,12.50,7.51,3.00,30.51,10.51,49.96,39.47
std,0.58,3.45,8.80,6.92,17.31,17.33,16.73,28.89,14.43,0.58,"1,443.46",288.86,103.67,1.41,7.60,56.78,370.13,226.72,161.93,28.89,7.22,4.03,1.41,17.03,5.77,28.85,10.94
min,"2,024.00",1.00,1.00,0.00,0.00,0.00,18.00,0.00,1.00,3.00,0.00,0.00,10.00,1.00,0.00,0.00,7.53,4.03,0.70,0.43,0.00,1.00,1.00,1.00,1.00,0.00,6.66
25%,"2,024.00",4.00,8.00,5.00,14.00,14.00,32.00,24.90,13.00,3.50,"1,253.00",250.00,69.51,2.00,0.00,0.00,139.02,82.45,50.69,10.12,6.26,4.00,2.00,15.70,6.00,25.00,31.69
50%,"2,025.00",7.00,16.00,11.00,30.00,29.00,46.00,50.00,25.00,4.00,"2,503.00",500.00,125.11,3.00,10.00,16.91,290.19,172.53,108.79,21.28,12.50,8.00,3.00,30.50,11.00,50.00,39.90
75%,"2,025.00",10.00,23.00,17.00,44.00,44.00,61.00,75.00,38.00,4.50,"3,752.00",750.00,195.61,4.00,15.00,49.26,545.76,326.45,214.64,40.59,18.74,11.00,4.00,45.30,16.00,74.90,48.17
max,"2,026.00",12.00,31.00,23.00,59.00,59.00,75.00,100.00,50.00,5.00,"5,000.00","1,000.00",500.00,5.00,25.00,624.71,"2,499.75","1,743.70","1,495.75",245.59,25.00,14.00,5.00,60.00,20.00,100.00,60.02


## 4. Missing Values

In [8]:
# Percentage of missing values per column (%)
print((ecommerce_df.isnull().sum() / ecommerce_df.shape[0] * 100).to_string())

order_id                    0.00
order_date                  0.00
order_year                  0.00
order_month                 0.00
order_day                   0.00
order_hour                  0.00
order_minute                0.00
order_second                0.00
is_weekend                  0.00
order_status                0.00
return_reason              90.01
customer_id                 0.00
customer_name               0.00
gender                      0.00
age                         0.00
customer_segment            0.00
country                     0.00
city                        0.00
customer_loyalty_score      0.00
total_orders_by_customer    0.00
account_creation_date       0.00
product_id                  0.00
product_name                0.00
category                    0.00
sub_category                0.00
brand                       0.00
product_rating_avg          0.00
product_reviews_count       0.00
stock_quantity              0.00
unit_price_usd              0.00
quantity  

**Notable missing values:**
- `return_reason` → ~90.01% missing (expected, only present for returned orders)
- `customer_feedback` → ~19.96% missing
- `coupon_code` → ~50.00% missing (expected, only present when a coupon was used)

## 5. Duplicate Check

In [9]:
# Total number of fully duplicate rows
duplicates_row = ecommerce_df.duplicated().sum()
print(f'Total Duplicate Rows : {duplicates_row}')

Total Duplicate Rows : 0


## 6. Unique Identifier Check

In [10]:
"""
    Check which columns have completely unique values
    True = potential unique identifier
 """
print(ecommerce_df.apply(lambda col: col.is_unique).to_string())

order_id                    False
order_date                   True
order_year                  False
order_month                 False
order_day                   False
order_hour                  False
order_minute                False
order_second                False
is_weekend                  False
order_status                False
return_reason               False
customer_id                 False
customer_name               False
gender                      False
age                         False
customer_segment            False
country                     False
city                        False
customer_loyalty_score      False
total_orders_by_customer    False
account_creation_date       False
product_id                  False
product_name                False
category                    False
sub_category                False
brand                       False
product_rating_avg          False
product_reviews_count       False
stock_quantity              False
unit_price_usd

> **Observation:**  
> The dataset does **not** have a reliable unique identifier column.  
> Although `order_date` appears unique in this sample, timestamps are not dependable as primary keys in real-world scenarios (collisions can occur).